### ISSUE TITLE
Suggestion: solve listing with websockets

### ISSUE THREAD

**Issue by bobvanluijt** [semi-technologies/weaviate#898]
Currently, it is not possible to efficiently list items from the graph. This is logical and understandable from the perspective of how we store data. However, because we also store c11y based data and the class name it might allow us to list with a few criteria based on this. Also, when creating a data-dump, this might come in handy.

Because the graph might get very big, it might be interesting to investigate if we can support listing through a WebSocket. Queries we could support are;
1. Show me all concepts with class-x. Or;
2. Show me all concepts based on these keywords (Explore{} search). Or;
3. Combination of the above.

When somebody uses this functionality, the data streams out of Weavite.

**Comment by etiennedi** [comment]
I don't think web sockets solve the underlying problem. HTTP connections (or other streaming mechanisms) aren't the problem. The problem is that with the NoSQL stack there is no inherent order to the items, i.e. we have no way of getting all data out other than with something like a Spark Job. The current database stack simply doesn't have a mechanism to data dump or produce long lists. The problem isn't how we get it to the user (http vs http2 streaming vs web sockets, etc.). The problem is that we can't get the data out of the dbs in this fashion.

If we want a data dump functionality (or similiar "long list" features), the only two options I see are:
1. Build a connector that supports this (i.e. probably a SQL-based on)
2. Use Spark to create an asynchronous data dump. This could also be saved directly into a GCS/S3 storage bucket, where it can be conveniently downloaded by the user.

We can investigate if there is a way dump the entire index (either the janus based one or the custom c11y index) and then manually page over those. As we speak, I'm unaware of whether ES offers such a functionality, but this I can find out. However, this would also only work with medium sized graphs. If the size of the index alone is larger than the memory available to a regular pod, this becomes unfeasible. This might be overcome by deploying a special "exporter" with a temporary file system. However, at that point it might be cheaper again to simply go the Spark route.

In summary, I think it might be possible to build an exporter, but the cost will be high. Most likely all processes will have to happen as async background tasks. Web Sockets are surely possible, but I don't think the current problem with large dumps is at the transport layer, but at the persistence layer, so they also don't necessarily help.

**Comment by etiennedi** [comment]
Elasticsearch has indeed a method to dump all indexed items, using scrolls. However, similarly to a spark job, they are not indended for real time use, so this would indeed go the route of using a specific "exporter" deployment: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-scroll.html

Nevertheless, a specialized tool like this makes sense in my opinion, as well send millions upon millions of read requests (one per item) to the db anyway, so a dump operation should happen over-night or in a designated maintenance window as it'd put massive load on the db.

**Comment by bobvanluijt** [comment]
Thanks for the insights @etiennedi. 

The use case here would be the creation of graph dumps or backups. 

Related to your remark: 

> Nevertheless, a specialized tool like this makes sense in my opinion, as we'll send millions upon millions of read requests (one per item) to the db anyway, so a dump operation should happen over-night or in a designated maintenance window as it'd put massive load on the db.

This is why I was suggesting something like a WebSocket. We could add super simple filtering when dumping. For example: `only of class X` or `where int property Y should be < 1000`, etc.

**Comment by etiennedi** [comment]
> The use case here would be the creation of graph dumps or backups.

For backups, I'd suggest to simply do a backup of the databases. That's a very straightforward process, as it's essentially just making snapshots of hard drives. No need to go through weaviate for that. 

Regarding a dump functionality, I can make a draft, but this is not prioritized very highly at the moment. If this becomes more urgent, please let me know.

> This is why I was suggesting something like a WebSocket.

I'm not sure if websockets would actually help us. In fact they can be [quite problematic on Kubernetes](https://stackoverflow.com/a/52346035). The issue we need to solve is the orchestration of how to get the stuff out of the index/database. Once we have that, we could simply stream into a GCS bucket for example. Happy to build a simple prototype, but as mentioned, I'm currently not prioritizing this. Please let me know once I should.

**Comment by bobvanluijt** [comment]
Cool, would love to hear thoughts and suggestions. 

@fefi42 feel free to weigh in on this discussion.

**Comment by fefi42** [comment]
I really have too little insight into the inner workings yet to say something about this issue. I just know that many traditional databases use compression e.g. wordbooks to get more data into memory. This might be an solution to make more data processable. Don't know if this helps.
